src/core/reference/include/openvino/reference/paged_attention.hpp

```diff
@@ -0,0 +1,330 @@
+// Copyright (C) 2018-2025 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include "openvino/core/shape.hpp"
+#include "openvino/core/type.hpp"
+#include "openvino/core/type/element_type.hpp"
+#include "openvino/op/paged_attention.hpp"
+#include "openvino/op/util/attr_types.hpp"
+
+namespace ov {
+namespace reference {
+namespace paged_attention_utils {
+
+// --- Helper / Unit Functions ---
+
+template <typename T>
+T dot_product(const T* a, const T* b, int32_t size) {
+    return std::inner_product(a, a + size, b, T(0));
+}
+
+template <typename T>
+void softmax(std::vector<T>& scores) {
+    T max_score = *std::max_element(scores.begin(), scores.end());
+    T sum = T(0);
+    for (auto& s : scores) {
+        s = std::exp(s - max_score);
+        sum += s;
+    }
+    for (auto& s : scores)
+        s /= sum;
+}
+
+/*
+ * Apply RoPE (Rotary Positional Embedding) rotation to a vector.
+ */
+template <typename T>
+void apply_rope(T* vec, int32_t head_size, const T* rotation_trig_lut, int32_t trig_index) {
+    int32_t half = head_size / 2;
+    const float* row = reinterpret_cast<const float*>(rotation_trig_lut) + trig_index * head_size;
```

Comment:
Why reinterpret cast here? If the `T` is e.g. 2-byte type the risk go outside buffer is big.

Reply:
Right, removed